tests/Unit/Http/UrlsTest.php

```diff
@@ -7,24 +7,109 @@
 
 final class UrlsTest extends BaseTestCase
 {
-    public function testSimpleUrlsAreReturned()
+
+    /**
+     * @dataProvider dataProviderSanitize
+     * @param string $url
+     * @param string $expected
+     * @return void
+     */
+    public function testSanitize($url, $expected)
     {
-        $this->assertSame('some_url.com/path/', Urls::sanitize('some_url.com/path/'));
+        /* This test is an exact replica of the same method in tests/Unit/private/UriTest.php and has to be kept in sync
+         * until the other test will be removed as part of the PHP->C migration
+         */
+        $this->assertSame($expected, Urls::sanitize($url));
     }
 
-    public function testSimpleUrlsWithSchemaAreReturned()
+    public function dataProviderSanitize()
     {
-        $this->assertSame('https://some_url.com/path/', Urls::sanitize('https://some_url.com/path/'));
+        return [
+            // empty
+            [null, ''],
+            ['', ''],
+
+            // with schema
+            ['https://some_url.com/path/', 'https://some_url.com/path/'],
+
+            // with no schema
+            ['some_url.com/path/', 'some_url.com/path/'],
+
+            // query and fragment
+            ['some_url.com/path/?some=value', 'some_url.com/path/'],
+            ['some_url.com/path/?some=value#fragment', 'some_url.com/path/'],
+
+            // userinfo
+            ['my_user:my_password@some_url.com/path/', '?:?@some_url.com/path/'],
+            ['my_user:@some_url.com/path/', '?:@some_url.com/path/'],
+            ['my_user:@some_url.com/path/?key=value', '?:@some_url.com/path/'],
+            ['https://my_user:my_password@some_url.com/path/', 'https://?:?@some_url.com/path/'],
+            ['https://my_user:@some_url.com/path/', 'https://?:@some_url.com/path/'],
+            ['https://my_user:@some_url.com/path/?key=value', 'https://?:@some_url.com/path/'],
+
+            // idempotency
+            ['https://?:@some_url.com/path/?key=value', 'https://?:@some_url.com/path/'],
+            ['?:?@some_url.com/path/?some=value#fragment', '?:?@some_url.com/path/'],
```

Comment:
I you truly want idempotency, also add a test without ?query_string part - because the strstr will then skip remove everything starting with the ? in the password.
That's why my regex suggestion wasn't attempting to be idempotent at all, because it hasn't been in the first place.

To be precise, we'd need to strip everything from the _first_ question mark _after_ any potential @.

```suggestion
            // idempotency
            ['https://?:@some_url.com/path/?key=value', 'https://?:@some_url.com/path/'],
            ['?:?@some_url.com/path/', '?:?@some_url.com/path/'],
            ['?:@some_url.com/path/?some=?#fragment', '?:@some_url.com/path/'],
```

better test cases

Reply:
Thanks @bwoebi for suggesting the test case. I think that your concern should be addressed by using `<sanitized> in place of `?` if my understanding is correct. But please let me know if I am missing something or even if you have a test case in addition to the one you added above for me to add.